### PR TITLE
Increase test coverage, improve docstrings, minor bug fixes

### PR DESCRIPTION
**Pull request by rfrenchseti**
# Overview

This PR makes a massive improvement to test coverage (from 73% to 97%), clarifies docstrings, and fixes a number of minor bugs that were found along the way.

The new test coverage is:

```
Name                                 Stmts   Miss Branch BrPart  Cover
----------------------------------------------------------------------
polymath/__init__.py                    15      0      0      0   100%
polymath/_version.py                     9      0      0      0   100%
polymath/boolean.py                     95      0     14      0   100%
polymath/extensions/__init__.py        167      0      0      0   100%
polymath/extensions/broadcaster.py      77      0     48      0   100%
polymath/extensions/indexer.py         308     11    196     11    95%
polymath/extensions/item_ops.py        134      0     30      0   100%
polymath/extensions/iterator.py         36      0      6      0   100%
polymath/extensions/mask_ops.py        158      0     62      0   100%
polymath/extensions/math_ops.py        803     50    382     21    93%
polymath/extensions/pickler.py         431     22    196     15    93%
polymath/extensions/shaper.py          186      0    114      0   100%
polymath/extensions/shrinker.py         86      7     48      5    91%
polymath/extensions/tvl.py             117      0     50      0   100%
polymath/extensions/vector_ops.py      266      0    132      0   100%
polymath/matrix3.py                    298      0     60      0   100%
polymath/matrix.py                     185      0     66      0   100%
polymath/pair.py                       105      0     44      0   100%
polymath/polynomial.py                 334      6    150      9    97%
polymath/quaternion.py                 356      0    104      0   100%
polymath/qube.py                      1283     31    718     30    97%
polymath/scalar.py                     778     28    402     20    96%
polymath/unit.py                       492      0    218      0   100%
polymath/vector3.py                    127      0     26      0   100%
polymath/vector.py                     371      7    168      7    97%
----------------------------------------------------------------------
TOTAL                                 7217    162   3234    118    97%
```

**All of the changes need to be carefully reviewed, especially the changes to the code to fix apparent bugs. In some cases the intent of the code was not clear, and in other cases it just appeared to be incorrect. It is entirely possible that some of the fixes are not correct!**

### `extensions/__init__.py`

- `__matmul__` was missing from the copying operators, meaning that the `@` operator could never work.

### `extensions/item_ops.py`

- Clarified docstrings.
- Fixed copy-paste errors in exceptions.

### `extensions/mask_ops.py`

- In `mask_where`, the docstring description of `recursive` does not agree with the normal meaning throughout the rest of the code. Please verify the new meaning and the code changes to make it work.

### `extensions/math_ops.py`

- When converting arguments to Qubes, there is a group of methods that converts using `as_this_type` and a group that uses `as_scalar`. This code hasn't changed (although the docstrings have been clarified). Please make sure this is the desired behavior.
- __rsub__ had a bug where if passed a Qube it would return None. This looks like it was an indentation error.

> The `__rsub__` method (right-side subtraction, implementing `arg - self`) had a bug where it would implicitly return
> `None` when the `arg` parameter was already a `Qube` object. The method had an `if` statement to handle the case
> when `arg` is not a `Qube`, but when `arg` was already a `Qube`, the function would fall through without an explicit
> return statement.

- Docstring cleanup.

### `extensions/pickler.py`

- As a general hint, if flake8 says a variable is unused but you're sure it's used in an error message, make sure you have the preceding `f`. This happened several time in the code base.
- Docstring cleanup.

### `extensions/shrinker.py`

- Docstring cleanup.

### `extensions/tvl.py`

- `tvl_and` and `tvl_or` did not fully document the truth table. The actual code seemed to have some subtle errors in it with masks. Please check the rewritten versions carefully.

> The docstring stated:
> > "False and anything = False (even if False is masked or the other operand is masked)"
> However, the correct interpretation (as clarified by the user) should be:
> > "False (unmasked) and anything = False (even if the other operand is masked)"
> The original implementation checked for False values regardless of whether they were masked, which was incorrect.

> Similar to `tvl_and`, the docstring stated:
> > "True or anything = True (even if True is masked or the other operand is masked)"
> The correct interpretation should be:
> > "True (unmasked) or anything = True (even if the other operand is masked)"
> The original implementation checked for True values regardless of mask.

- Docstring cleanup.

> `tvl_and` and `tvl_or` - The docstrings stated:
> > "For array operations, the rules apply element-wise. When operands are masked in array contexts, the result may be masked even when the simple scalar rules would suggest a definite True/False result."
> This suggested that arrays had different behavior than scalars, which was incorrect.

### `extensions/vector_ops.py`

- The code at line 61 appears to be unreachable to I replaced it with an exception to catch any future changes.
- The code at line 140 appears to display the wrong values.
- The code at line 167 was nonsensical from what appears to be an indent error.
- I made `cross_2x2` and `cross_3x3` internal functions.

### `matrix.py`

- Bugs in `from_scalar`.
- Removed Python 2 support.
- Docstring cleanup.

### `matrix3.py`

- `poly_rotation` had a bad zero argument to `np.stack`.

> The `pole_rotation()` method failed when given array inputs for `ra` (right ascension) or `dec` (declination) due to a scalar literal `0.` in the `np.stack()` call that didn't broadcast correctly with array-shaped trigonometric values.
> **Original Code**:
> ```python
> values = np.stack([-sin_ra,            cos_ra,           0.,
>                    -cos_ra * sin_dec, -sin_ra * sin_dec, cos_dec,
>                     cos_ra * cos_dec,  sin_ra * cos_dec, sin_dec],
>                   axis=-1)
> ```
> **Error**:
> ```
> ValueError: all input arrays must have the same shape
> ```
> This occurred when `ra` or `dec` were arrays (e.g., `np.array([0., np.pi/4])`), causing `sin_ra` and `cos_ra` to be arrays, while `0.` remained a scalar. NumPy's `np.stack()` requires all inputs to have the same shape.

- `__imul__` called `_set_values` with a bad `example` argument.
- `from_euler` and `to_euler` did not work with a tuple of axes because `lower` was called on the tuple before it was detected.
- Removed Python 2 support.
- Docstring cleanup.

### `pair.py`

- `from_scalars` claimed to support `None` arguments but didn't.

 >The original `Pair.from_scalars()` method did not handle `None` values for the `x` and `y` parameters. When `None` was > passed, the code would fail with:
> ```
> TypeError: invalid Scalar data type: <class 'NoneType'>
> ```
> This occurred because `Qube.from_scalars()` (which was called directly) attempted to convert `None` to a `Scalar`, which is not a valid conversion.

- `rot90` needed to pass `recursive` as a named argument.

> The `rot90()` method had a bug when handling derivatives. The code attempted to call:
> ```python
> obj.insert_deriv(key, deriv.rot90(False))
> ```
> However, `rot90()` has a keyword-only parameter `recursive`, so passing `False` as a positional argument caused:
> ```
> TypeError: Pair.rot90() takes 1 positional argument but 2 were given
> ```

- Removed Python 2 support.
- Docstring cleanup.

### `polynomial.py`

- The handling of derivatives in `invert_line` is highly questionable and needs to be validated.
- For `__iadd__` and `__isub__` this is not valid Python: `self.super().__iadd__(arg.super())`. Cursor wrote a fix to handle polynomials of different order, but it needs to be validated and there's probably a way to simplify it.

> The in-place addition and subtraction methods (`__iadd__` and `__isub__`) were using the anti-pattern `self = ...`, which creates a new local variable instead of modifying the existing object. This violates the contract of in-place operations.

> The code called `self._add_derivs(self, arg)` and `self._sub_derivs(self, arg)`, but these are class methods that expect two arguments. The correct call should use `Qube._add_derivs(self, arg)` and `Qube._sub_derivs(self, arg)`.

> The code contained an unreachable branch in the conditional logic for handling polynomial order alignment. The `else` branch (`arg.set_order(max_order)`) was mathematically impossible to execute.

- Likewise in `__sub__` and `__rsub__` this is not valid Python: `return Polynomial(self.super() - arg.super())` and was replaced with `Polynomial(self.as_vector() - arg.as_vector())`.
- `__mul__` had some problems and was rewritten.
- `__pow__` did not work at all and was rewritten.
- `__eval__` and `roots` had a lot of issues and were significantly changed.

> The `eval()` method had a bug when extracting constant values from polynomials with denominators (drank > 0). The > original code used:
> ```python
> const_values = self._values[(Ellipsis, 0) + tail_indx]
> ```
> This failed because `tail_indx` is a tuple of slices, and tuple concatenation with `+` doesn't work correctly in this context > when used for NumPy indexing.

> For zero-order (constant) polynomials, the code attempted to create a `Scalar` using `Scalar(example=self)` without providing the required `arg` parameter. The `Scalar` constructor requires an `arg` parameter as the first positional argument.

> When `total_shifts` is an array (not a scalar), the code attempted to use boolean indexing incorrectly: `root_mask[total_shifts > k, k]`. This fails because the boolean array shape doesn't match the root_mask shape for advanced indexing.

> The duplicate detection logic was applied after calling `roots.sort()`, which converts masked values to `inf`. This makes it impossible to detect duplicates because masked values are no longer comparable.

> The code called `self.deriv.eval(...)` but `deriv` is a method, not a property, so it needs to be called first: `self.deriv().eval(...)`.

> In the scalar case of the `while np.any(shifts)` loop, the code was using `break` after shifting once, which would exit the loop even if the leading coefficient was still zero. This could leave leading zeros in the `coefficients` array, causing `np.linalg.eigvals` to fail.

- Docstring cleanup.

### `quaternion.py`

- `to_matrix3` didn't support scalar pnorms.

> The code attempted to handle scalar `pnorm` values (when the quaternion has shape `()`) but had two issues:
> 1. It checked `np.shape(pvals) == ()` instead of `np.shape(pnorm) == ()` when determining if `pnorm` was scalar
> 2. When `pnorm` was scalar, the code tried to use it directly in array operations with `pnorm[..., np.newaxis]`, which would fail because scalar NumPy values don't support indexing

- `from_euler` did not work with a tuple of axes because `lower` was called on the tuple before it was detected.
- Removed Python 2 support.
- Docstring cleanup.

### `qube.py`

- `_suitable_dtype` missing `f` string prefix.
- `delete_derives` modified a dict that was being iterated over.
- `without_unit` didn't remove units from derivatives.

> The docstring stated:
> > "If recursive is True, derivatives are also stripped of their units."
> However, the implementation did not actually strip units from derivatives when `recursive=True`.

- `as_float` called `__copy__` with a `recursive` parameter but that parameter is not supported by `__copy__`.
- `as_size_zero` was not operating properly with scalars.

> Fixed bug in polymath/qube.py line 2621: When as_size_zero is called with axis=0 and the mask is a scalar, the mask
> shape was (0,) but should match the values shape (0, 2). Fixed by creating a mask array that matches the shape of
> new_values.

- Docstring cleanup.

> `into_unit` - The docstring stated:
> > "Converted to its unit"
> This was ambiguous about the direction of conversion - does it convert TO the unit or FROM the unit?

### `scalar.py`

- `sort` did not properly replace masked values with the maximum.

> The code attempted to assign `result.max()` directly to an array slice, but `result.max()` returns a `Scalar` object, not a raw NumPy value.

### `unit.py`

- `mul_units` and `div_units`. I really did not understand these methods. There is an argument `name` that was ignored entirely. If a new unit was created, its name was set to `None`, but if one of the original units was returned, it kept its name. I cleaned up the logic and made it set the new unit's name to `name`, but the overall intent of this function needs to be verified.
- `name_to_dict` wraps parsing in at 99,999 characters in a way that can't possibly work. I don't understand the purpose of the wrap. If it's a limit, we should just throw an exception. This code needs to be verified.
- Docstring cleanup.

### `vector.py`

- Removed Python 2 support.
- Docstring cleanup.

### `vector3.py`

- `from_scalars` claimed to support `None` arguments but didn't. Same as `pair.py`.
- Removed Python 2 support.
- Docstring cleanup.

### `pyproject.toml`

- Removed support for Python 3.8 and 3.9.

## New Test Files

- There are roughly 16,000 lines of new test cases.
- Because some of these test files were produced from scratch without regards to what was already being tested, it is likely there are multiple tests that cover the same functions and lines. But I don't think this is a problem. But tests still run in about 13 seconds and more ways of testing a function is always good.

<hr>

## Summary by CodeRabbit

* **Python Version Update**
  * Minimum Python requirement raised to 3.10.

* **New Features**
  * Matrix multiplication via matmul for qube-like objects.
  * Euler-angle inputs for quaternions accept tuple/list axes as well as strings.

* **Public API**
  * into_unit now enforces keyword-only usage (recursive keyword).

* **Documentation**
  * Extensive docstring clarifications across core numeric types and operations.

* **Bug Fixes / Behavior**
  * Tighter shape validation, improved masking, derivative propagation, unit handling, and sort behavior.

* **Tests**
  * Large expansion of unit tests covering many classes and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: polymath/polynomial.py [hunk 1]
```diff
@@ -607,26 +796,42 @@ def roots(self, recursive=True):
...
+                        # root_mask has shape (order, ...shape...)
+                        # mask_indices gives indices into the shape dimensions
+                        # We need to prepend k for the first dimension
+                        root_mask[(k,) + mask_indices] = True
+        else:
+            # Scalar case
+            if total_shifts > 0:
+                for k in range(int(total_shifts)):
+                    root_mask[k, ...] = True
+
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply polymath/polynomial.py:819]: I don't understand this comment. The if is based on total_shifts, which can be a scalar, not on root_mask.

File: tests/test_scalar_coverage.py [hunk 2]
```diff
@@ -0,0 +1,1166 @@
...
+        a = Scalar(1.)
+        Qube.prefer_builtins(True)
+        b = a.sign()
+        # sign() returns the sign, which for float 1.0 is 1.0 (float), not int
+        # But if it's an integer Scalar, it might return int
+        a_int = Scalar(1)  # Integer
+        b_int = a_int.sign()
+        # The result type depends on the input type
+        self.assertIsInstance(b, (int, float))
+        Qube.prefer_builtins(False)
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_scalar_coverage.py]: b_int is used

[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_scalar_coverage.py]: It's literally right there:             self.assertIsInstance(b_int, int)

[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_scalar_coverage.py]: Yes I haven't pushed it yet

File: tests/test_math_ops_coverage.py [hunk 3]
```diff
@@ -0,0 +1,854 @@
...
+
+        ##################################################################################
+        # Test __rmul__ error cases
+        ##################################################################################
+        # Test exception revision (line 452-455)
+        try:
+            a = Scalar([1., 2., 3.])
+            _ = object().__rmul__(a)  # This won't work, but tests the path
+        except (TypeError, AttributeError):
+            pass
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_math_ops_coverage.py]: These are covered in other files

File: tests/test_vector3_basic.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+        # Test as_vector3 with n-D 1x3 Matrix
+        m1x3_nd = Matrix([[[1., 2., 3.]], [[4., 5., 6.]]])
+        self.assertEqual(m1x3_nd.shape, (2,))
+        self.assertEqual(m1x3_nd._numer, (1, 3))
+        v1x3_nd_conv = Vector3.as_vector3(m1x3_nd)
+        self.assertEqual(type(v1x3_nd_conv), Vector3)
+        self.assertEqual(v1x3_nd_conv.shape, (2,))
+        self.assertTrue(np.allclose(v1x3_nd_conv.vals[0], [1., 2., 3.]))
+        self.assertTrue(np.allclose(v1x3_nd_conv.vals[1], [4., 5., 6.]))
+
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_vector3_basic.py:139]: Vector3 has 100% test coverage when the tests are taken as a whole

File: tests/test_qube_coverage.py [hunk 5]
```diff
@@ -0,0 +1,1939 @@
...
+        # When mask is scalar and antimask is provided, the mask needs to be expanded
+        # The code at line 1163-1167 handles this by expanding the mask
+        a._set_values(new_values, mask=True, antimask=antimask)
+        # After expansion, mask should be an array
+        # Only elements where antimask is True get set to True
+        self.assertIsInstance(a.mask, np.ndarray)
+        self.assertTrue(a.mask[0])
+        self.assertFalse(a.mask[1])  # antimask[1] is False, so mask[1] stays False
+        self.assertTrue(a.mask[2])
+
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_qube_coverage.py:1775]: They are not identical because the masks are different

File: tests/test_boolean.py [hunk 6]
```diff
@@ -800,4 +800,69 @@ def runTest(self):
...
+        a = Boolean([True, False])
+        result = 2.0 / a
+        self.assertEqual(result[0], 2.0)
+        self.assertEqual(result[1], Scalar.MASKED)
+
+        # Test __rtruediv__ with Qube arg
+        a = Boolean([True, False])
+        result = a / a
+        self.assertEqual(result[0], 1.0)
+        self.assertEqual(result[1], Scalar.MASKED)
```
[the remark this answers is not in the document]
Comment (rfrenchseti) [reply tests/test_boolean.py:823]: Empirically this is not true. Code coverage is different in __rtruediv__ with this test enabled.

**Comment by rfrenchseti** [comment]
@coderabbit summary

**Comment by rfrenchseti** [comment]
@coderabbit summary
